R/zchunk_LB172.nonghg_elc_USA.R

```diff
@@ -0,0 +1,165 @@
+      filter(Non.CO2 == "SO2") %>%
+      mutate(Non.CO2 = "SO2_ELEC")
+    # Add these sudo species into L172 main data
+    L172.nonghg_tg_state_elec_F_Yb <- bind_rows(L172.nonghg_tg_state_elec_F_Yb,
+                                                L172.NOx_Coal_Yb, L172.SO2_Coal_Yb,
+                                                L172.NOx_ELEC_Yb, L172.SO2_ELEC_Yb)
+
+
+    ## Input-based emissions in the future years, use data from EPA-ORD
+    # L172.nonghg_tgej_state_elec_F_Yf: Electricity non-co2 emissions coefficients by fuel input and U.S. state in future model years
+
+    L172.nonghg_tgej_state_elec_F_Yf_prior2025 <- EPA_state_egu_emission_factors_ktPJ %>%
+      # Convert to long format
+      gather(variable, value, -state_name, -fuel) %>%
+      separate(variable, into = c("year","Non.CO2"), sep = "_") %>%
+      mutate(year = as.numeric(year)) %>%
+      # NOTE: for now change oil to refined liquids
+      mutate(fuel = gsub("oil","refined liquids", fuel)) %>%
+      # state code & select relevant columns
+      left_join_error_no_match(states_subregions %>% select(state, state_name),
+                by = c("state_name")) %>%
+      mutate(sector = "elec_heat") %>%
+      select(state, sector, fuel, Non.CO2, year, value) %>%
+      # filter only future model years prior to 2025
+      filter(year %in% FUTURE_YEARS[1:3])
+
+    # Add remaining future years, keeping them constant at 2025 levels
+    L172.nonghg_tgej_state_elec_F_Yf_post2025 <- L172.nonghg_tgej_state_elec_F_Yf_prior2025 %>%
+      filter(year == 2025) %>%
+      select(-year) %>%
+      repeat_add_columns(tibble("year" =  FUTURE_YEARS[4:18]))
+
+    # combine together
+    L172.nonghg_tgej_state_elec_F_Yf <- bind_rows(L172.nonghg_tgej_state_elec_F_Yf_prior2025,
+                                              L172.nonghg_tgej_state_elec_F_Yf_post2025) %>%
+      na.omit
+
+    ## Add sodu Non.CO2 species for NOx_Coal, SO2_Coal, NOx_ELEC, SO2_ELEC for future years
+    # 1) Add N0x_Coal
+    L172.NOx_Coal_Yf <- L172.nonghg_tgej_state_elec_F_Yf %>%
+      filter(fuel == "coal" & Non.CO2 == "NOx") %>%
+      mutate(Non.CO2 = "NOx_Coal")
+    # 2) Add S02_Coal
+    L172.SO2_Coal_Yf <- L172.nonghg_tgej_state_elec_F_Yf %>%
+      filter(fuel == "coal" & Non.CO2 == "SO2") %>%
+      mutate(Non.CO2 = "SO2_Coal")
+    # 3) Add NOx_ELEC
+    L172.NOx_ELEC_Yf <- L172.nonghg_tgej_state_elec_F_Yf %>%
+      filter(Non.CO2 == "NOx") %>%
+      mutate(Non.CO2 = "NOx_ELEC")
+    # 4) Add SO2_ELEC
+    L172.SO2_ELEC_Yf <- L172.nonghg_tgej_state_elec_F_Yf %>%
+      filter(Non.CO2 == "SO2") %>%
+      mutate(Non.CO2 = "SO2_ELEC")
+    # Add these sudo species into L172 main data
+    L172.nonghg_tgej_state_elec_F_Yf <- bind_rows(L172.nonghg_tgej_state_elec_F_Yf,
+                                                  L172.NOx_Coal_Yf, L172.SO2_Coal_Yf,
+                                                  L172.NOx_ELEC_Yf, L172.SO2_ELEC_Yf)
+
```

Comment:
This whole section repeats lines 70-87 right? In that case let's use a function or otherwise reduce redundancy.

Reply:
the entire data system only needs these two places to have these codes, even it's a little bit redundant, seems not a huge waste of resource to repeat. I'm not sure how to create a helper function with such specific purpose.